// ngx-page-scroll/src/ngx-page-scroll.directive.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Directive, Input, Output, EventEmitter, Inject, Optional } from '@angular/core';
import { Router, NavigationEnd, NavigationError, NavigationCancel } from '@angular/router';
import { DOCUMENT } from '@angular/common';
import { PageScrollService } from './ngx-page-scroll.service';
import { PageScrollInstance } from './ngx-page-scroll-instance';
import { PageScrollUtilService as Util } from './ngx-page-scroll-util.service';
import { EasingLogic } from './ngx-page-scroll-config';
var PageScroll = /** @class */ (function () {
    function PageScroll(pageScrollService, router, document) {
        this.pageScrollService = pageScrollService;
        this.router = router;
        this.pageScrollTarget = null;
        this.pageScrollHorizontal = null;
        this.pageScrollOffset = null;
        this.pageScrollDuration = null;
        this.pageScrollSpeed = null;
        this.pageScrollEasing = null;
        this.pageScrollAdjustHash = false;
        this.pageScroll = null;
        this.pageScrollFinish = new EventEmitter();
        this.document = document;
    }
    PageScroll.prototype.ngOnChanges = function (changes) {
        // Some inputs changed, reset the pageScrollInstance
        this.pageScrollInstance = undefined;
    };
    PageScroll.prototype.ngOnDestroy = function () {
        if (this.pageScrollInstance) {
            this.pageScrollService.stop(this.pageScrollInstance);
        }
        return undefined;
    };
    PageScroll.prototype.generatePageScrollInstance = function () {
        if (Util.isUndefinedOrNull(this.pageScrollInstance)) {
            this.pageScrollInstance = PageScrollInstance.newInstance({
                document: this.document,
                scrollTarget: this.pageScrollTarget || this.href,
                scrollingViews: null,
                namespace: this.pageScroll,
                verticalScrolling: !this.pageScrollHorizontal,
                pageScrollOffset: this.pageScrollOffset,
                pageScrollInterruptible: this.pageScrollInterruptible,
                pageScrollEasingLogic: this.pageScrollEasing,
                pageScrollDuration: this.pageScrollDuration,
                pageScrollSpeed: this.pageScrollSpeed,
                pageScrollFinishListener: this.pageScrollFinish
            });
        }
        return this.pageScrollInstance;
    };
    PageScroll.prototype.pushRouterState = function () {
        if (this.pageScrollAdjustHash && typeof this.pageScrollInstance.scrollTarget === 'string'
            && this.pageScrollInstance.scrollTarget.substr(0, 1) === '#') {
            // "Navigate" to the current route again and this time set the fragment/hash
            this.router.navigate([], {
                fragment: this.pageScrollInstance.scrollTarget.substr(1),
                preserveQueryParams: true
            });
        }
    };
    PageScroll.prototype.scroll = function () {
        var pageScrollInstance = this.generatePageScrollInstance();
        this.pushRouterState();
        this.pageScrollService.start(pageScrollInstance);
    };
    PageScroll.prototype.handleClick = function (clickEvent) {
        var _this = this;
        if (this.routerLink && this.router !== null && this.router !== undefined) {
            var urlTree = void 0;
            if (typeof this.routerLink === 'string') {
                urlTree = this.router.parseUrl(this.routerLink);
            }
            else {
                urlTree = this.router.createUrlTree(this.routerLink);
            }
            if (!this.router.isActive(urlTree, true)) {
                // We need to navigate their first.
                // Navigation is handled by the routerLink directive
                // so we only need to listen for route change
                var subscription_1 = this.router.events.subscribe(function (routerEvent) {
                    if (routerEvent instanceof NavigationEnd) {
                        subscription_1.unsubscribe();
                        // use a timeout to start scrolling as soon as the stack is cleared
                        setTimeout(function () {
                            _this.scroll();
                        }, 0);
                    }
                    else if (routerEvent instanceof NavigationError || routerEvent instanceof NavigationCancel) {
                        subscription_1.unsubscribe();
                    }
                });
                return false; // to preventDefault()
            }
        }
        this.scroll();
        return false; // to preventDefault()
    };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], PageScroll.prototype, "routerLink", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], PageScroll.prototype, "href", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], PageScroll.prototype, "pageScrollTarget", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PageScroll.prototype, "pageScrollHorizontal", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], PageScroll.prototype, "pageScrollOffset", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], PageScroll.prototype, "pageScrollDuration", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], PageScroll.prototype, "pageScrollSpeed", void 0);
    __decorate([
        Input(),
        __metadata("design:type", EasingLogic)
    ], PageScroll.prototype, "pageScrollEasing", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], PageScroll.prototype, "pageScrollInterruptible", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], PageScroll.prototype, "pageScrollAdjustHash", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], PageScroll.prototype, "pageScroll", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], PageScroll.prototype, "pageScrollFinish", void 0);
    PageScroll = __decorate([
        Directive({
            selector: '[pageScroll]',
            host: {
                '(click)': 'handleClick($event)',
            }
        }),
        __param(1, Optional()), __param(2, Inject(DOCUMENT)),
        __metadata("design:paramtypes", [PageScrollService, Router, Object])
    ], PageScroll);
    return PageScroll;
}());
export { PageScroll };
